[expo-updates][android] remove UPDATES_CONFIGURATION_USES_LEGACY_MANIFEST_KEY constant

jkhales opened the pull request:
# Why

Remove UPDATES_CONFIGURATION_USES_LEGACY_MANIFEST_KEY to make setup easier. 

We check for `expo-protocol-version` instead and use the legacy manifest only if it is undefined. Legacy manifests are in some sense 'protocol version: `null`'

# How

There were 3 places that `useLegacyManifest` was called.

1. after downloading a manifest (no change needed)

However the other two were problematic as they did not assume an http response and thus a chance to check its headers:

2. when loading an embedded manifest (removed check, was unnecessary)
3. when making query headers (removed cache-control header and made it so that the server set these instead: https://github.com/expo/universe/pull/7121)

# Test Plan

Tested both new and legacy updates still work.

Deleted deprecated tests and added a new one for protocol error.

Commit:
[expo-updates] remove LEGACY_MANIFEST constant

Commit:
fix tests

jkhales commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/manifest/ManifestFactoryTest.java:
```diff
@@ -10,20 +11,20 @@
   private String bareManifestJson = "{\"id\":\"0eef8214-4833-4089-9dff-b4138a14f196\",\"commitTime\":1609975977832}";
 
-  private UpdatesConfiguration createConfig(boolean usesLegacyManifest) {
+  private UpdatesConfiguration createConfig() {
```
The useLegacyManifest variable is no longer set with `loadValuesFromMap`. For most of the tests the default `true` value is used. I moved it out of this helper function to keep this clear.

jkhales commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/manifest/ManifestFactoryTest.java:
```diff
@@ -10,20 +10,20 @@
 public class ManifestFactoryTest {
 
   private String legacyManifestJson = "{\"sdkVersion\":\"39.0.0\",\"id\":\"@esamelson/native-component-list\",\"releaseId\":\"0eef8214-4833-4089-9dff-b4138a14f196\",\"commitTime\":\"2020-11-11T00:17:54.797Z\",\"bundleUrl\":\"https://d1wp6m56sqw74a.cloudfront.net/%40esamelson%2Fnative-component-list%2F39.0.0%2F01c86fd863cfee878068eebd40f165df-39.0.0-ios.js\"}";
-  private String newManifestJson = "{\"manifest\":{\"runtimeVersion\":\"1\",\"id\":\"0eef8214-4833-4089-9dff-b4138a14f196\",\"createdAt\":\"2020-11-11T00:17:54.797Z\",\"launchAsset\":{\"url\":\"https://d1wp6m56sqw74a.cloudfront.net/%40esamelson%2Fnative-component-list%2F39.0.0%2F01c86fd863cfee878068eebd40f165df-39.0.0-ios.js\",\"contentType\":\"application/javascript\"}}}";
```
Removed this manifest top level key as it is deprecated.

esamelson commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/loader/FileDownloaderTest.java:
```diff
@@ -27,30 +27,6 @@ public void setup() {
     context = InstrumentationRegistry.getInstrumentation().getTargetContext();
   }
 
-  @Test
-  public void testCacheControl_LegacyManifest() {
-    HashMap<String, Object> configMap = new HashMap<>();
-    configMap.put("updateUrl", Uri.parse("https://exp.host/@test/test"));
-    configMap.put("runtimeVersion", "1.0");
-    configMap.put("usesLegacyManifest", true);
-    UpdatesConfiguration config = new UpdatesConfiguration().loadValuesFromMap(configMap);
-
-    Request actual = FileDownloader.setHeadersForManifestUrl(config, null, context);
-    Assert.assertEquals("no-cache", actual.header("Cache-Control"));
-  }
-
-  @Test
-  public void testCacheControl_NewManifest() {
-    HashMap<String, Object> configMap = new HashMap<>();
-    configMap.put("updateUrl", Uri.parse("https://exp.host/manifest/00000000-0000-0000-0000-000000000000"));
-    configMap.put("runtimeVersion", "1.0");
-    configMap.put("usesLegacyManifest", false);
-    UpdatesConfiguration config = new UpdatesConfiguration().loadValuesFromMap(configMap);
-
-    Request actual = FileDownloader.setHeadersForManifestUrl(config, null, context);
-    Assert.assertNull(actual.header("Cache-Control"));
-  }
```
let's keep both of these tests and just change the assertion in the first one to match the second (now neither request should have a `Cache-Control` header), since we don't want unintentional regressions on this behavior.

jkhales replied:
oh yea, good point.

esamelson commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/manifest/ManifestFactoryTest.java:
```diff
@@ -10,20 +10,20 @@
 
 import java.util.HashMap;
 
+import androidx.room.Update;
```
can remove this

esamelson commented on packages/expo-updates/android/src/main/java/expo/modules/updates/loader/FileDownloader.java:
```diff
@@ -118,6 +118,9 @@ public void onResponse(Call call, Response response) throws IOException {
 
           try {
+            String expoProtocolVersion = response.header("expo-protocol-version", null);
+            configuration.setUsesLegacyManifest(expoProtocolVersion == null);
```
IIRC we already pass the response into the `ManifestFactory` function -- could we just check the header and make the decision there? Would be nice to eliminate this key from the `UpdatesConfiguration` entirely and just rely on the response. (You might need to mock the response in the tests -- you can see an example of that in ManifestMetadataTest.java.)

jkhales replied:
ah ok, that makes sense. Let me try.

Commit:
feed back round 1

Commit:
revert accident

Commit:
refactor away from config

jkhales commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/manifest/ManifestFactoryTest.java:
```diff
@@ -13,72 +13,75 @@
     Assert.assertTrue(actual instanceof NewManifest);
   }
 
-  @Test
-  public void testGetEmbeddedManifest_Legacy() throws JSONException {
-    Manifest actual = ManifestFactory.getEmbeddedManifest(
-      new JSONObject(legacyManifestJson),
-      createConfig(true)
-    );
-    Assert.assertTrue(actual instanceof LegacyManifest);
-  }
```
this was not deleted, just moved below the new `testGetManifest_Error` that git thinks is replacing it.

jkhales commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/manifest/ManifestFactoryTest.java:
```diff
@@ -13,72 +13,75 @@
   }
 
   @Test
-  public void testGetEmbeddedManifest_New() throws JSONException {
```
The last two tests were indeed deleted.

esamelson commented on packages/expo-updates/android/src/main/java/expo/modules/updates/loader/FileDownloader.java:
```diff
@@ -312,11 +312,6 @@ private static Request setHeadersForUrl(Uri url, UpdatesConfiguration configurat
 
-    // legacy manifest loads should ignore cache-control headers from the server and always load remotely
-    if (configuration.usesLegacyManifest()) {
-      requestBuilder = requestBuilder.cacheControl(CacheControl.FORCE_NETWORK);
```
reminder that this PR should go under Breaking Changes in the changelog due to this change

esamelson commented on packages/expo-updates/android/src/main/java/expo/modules/updates/manifest/ManifestResponse.java:
```diff
@@ -15,6 +15,6 @@ public ManifestResponse(Response response) {
 
   public String header(String name) {
-    return mResponse.header(name);
+    return mResponse.header(name, null);
```
let's just additionally implement `public String header(String name, String defaultValue)` to keep the behavior of this wrapper as true to `okhttp3.Response` as we can

esamelson reviewed (changes requested):
looking good! just a few small changes, then I think we can land.

Commit:
fix typing

Commit:
Apply suggestions from code review

Co-authored-by: Eric Samelson <esamelson@users.noreply.github.com>

Commit:
fix tests

Commit:
add nullable and changelog

Commit:
better override

esamelson commented on packages/expo-updates/android/src/main/java/expo/modules/updates/manifest/ManifestFactory.java:
```diff
@@ -9,28 +9,23 @@
+    if (expoProtocolVersion == null) {
       return LegacyManifest.fromLegacyManifestJson(manifestJson, configuration);
-    } else {
+    } else if ( Integer.valueOf(expoProtocolVersion) == 0) {
```
```suggestion
    } else if (Integer.valueOf(expoProtocolVersion) == 0) {
```

esamelson commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/manifest/ManifestFactoryTest.java:
```diff
@@ -13,72 +13,74 @@
-  }
+  @Test(expected = Exception.class)
+  public void testGetManifest_UnsupportedProtocolVersion() throws Exception {
+    ManifestResponse response = mock(ManifestResponse.class);
```
should we mock `when(response.header("expo-protocol-version", null)).thenReturn("1");` here?

esamelson commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/manifest/ManifestFactoryTest.java:
```diff
@@ -13,72 +13,74 @@
-  public void testGetManifest_Legacy() throws JSONException {
+  public void testGetManifest_Legacy() throws Exception {
+    ManifestResponse response = mock(ManifestResponse.class);
+    when(response.header("expo-protocol-version")).thenReturn(null);
```
```suggestion
    when(response.header("expo-protocol-version", null)).thenReturn(null);
```

jkhales replied:
Ah, lol, great point. Technically we don't even need to mock this I guess, but explicit is better than implicit.

esamelson reviewed (changes requested):
sorry, just a few more comments still 😅

jkhales commented on packages/expo-updates/android/src/main/java/expo/modules/updates/manifest/ManifestResponse.java:
```diff
@@ -14,7 +15,13 @@ public ManifestResponse(Response response) {
+  @Nullable
   public String header(String name) {
-    return mResponse.header(name);
+    return this.header(name,null);
```
That's fair, I was thinking to keep them intentionally coupled as they are overloaded functions.

esamelson replied:
Your thinking makes sense and would be what I would do in almost any other instance of overloaded functions -- this is just a weird special case because this particular class only exists to wrap okhttp3.Response as transparently/dumbly as possible.

Commit:
Apply suggestions from code review

Co-authored-by: Eric Samelson <esamelson@users.noreply.github.com>

jkhales commented on packages/expo-updates/CHANGELOG.md:
```diff
@@ -17,6 +17,8 @@
 ### 🛠 Breaking changes
 
 - Dropped support for iOS 10.0 ([#11344](https://github.com/expo/expo/pull/11344) by [@tsapeta](https://github.com/tsapeta))
+- remove UPDATES_CONFIGURATION_USES_LEGACY_MANIFEST_KEY constant. ([#12181](https://github.com/expo/expo/pull/12181) by [@jkhales](https://github.com/jkhales))
```
Ah OK, so is the flow to always add to `Unpublished` and then when right before a publish we shift everything over?

Commit:
more feedback

jkhales commented:
> sorry, just a few more comments still 😅

appreciate the attention to detail! 🙏

esamelson reviewed (approved):
lgtm, thanks for addressing all of these comments!! as we are right in the middle of versioned QA it would be great if we could wait a few days to land this -- it's a lot easier if there are no breaking changes on master and we can just keep the sdk-41 branch up to date with master. is that ok with you? 😅

jkhales commented:
> lgtm, thanks for addressing all of these comments!! 

Thank you for making them!

> wait a few days to land this -- it's a lot easier if there are no breaking changes on master and we can just keep the sdk-41 branch up to date with master. is that ok with you? 😅

Definitely.
